Add editor onboarding tests

dcalhoun opened the pull request:
## Description
Add automated tests for existing native editor onboarding features.

This work also fixes a React Native warning caused by webpack-oriented Jest [image mock](https://git.io/JEFo0). The previous image mock was implemented following the Jest documentation [related to webpack](https://bit.ly/3jAqaJL). However, the `source` prop for React Native's [Image component](https://bit.ly/2YiuDbS) does not accept strings. The previous image mock resulted in the following warning in testing logs.

```
Warning: Failed prop type: Invalid prop `source` supplied to `Image`.
```

To address this, the previous image mock was removed in favor of relying upon the transform that is provided by the `react-native` [Jest preset](https://git.io/JEFor). In order to have this transform applied to [images within React Navigation](https://git.io/JEFoK), additional regex was added to `transformIgnorePatterns`.

## How has this been tested?
Verify native unit tests passed. 

## Screenshots <!-- if applicable -->
n/a

## Types of changes
Improve code quality. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
Add tooltip display message test

dcalhoun committed:
Add tap screen dismisses tooltip test

dcalhoun committed:
Add keyboard hide dismisses tooltip test

dcalhoun committed:
Simplify tooltip tests

dcalhoun committed:
Fix React Native warning caused by webpack-oriented Jest image mock

The previous image mock was implemented following the Jest documentation
related to webpack. However, the `source` prop for React Native's Image
component does not accept strings. The previous image mock resulted in
the following warning in testing logs.

```
Warning: Failed prop type: Invalid prop `source` supplied to `Image`.
```

To address this, the previous image mock was removed in favor of relying
upon the transform that is provided by the `react-native` Jest preset.
In order to have this transform applied to images within React
Navigation, additional regex was added to `transformIgnorePatterns`.

- https://git.io/JEFo0
- https://bit.ly/3jAqaJL
- https://bit.ly/2YiuDbS
- https://git.io/JEFor
- https://git.io/JEFoK

dcalhoun committed:
Add editor Help section tests

dcalhoun commented on packages/components/src/tooltip/index.native.js:
```diff
@@ -54,11 +55,12 @@ const useKeyboardVisibility = () => {
 			}
 		} );
 		return () => {
+			isCurrent = false;
 			showListener.remove();
 			hideListener.remove();
```
The removal of the listeners appears to not cancel execution of any currently registered callback. There were times where `setKeyboardVisible` as called on an unmounted component, resulting in React warnings. The addition of `isCurrent` prevents state updates to the component if it is no longer current.

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -68,6 +68,7 @@ function EditorHelpTopics( { close, isVisible, onClose } ) {
 			hideHeader
 			hasNavigation
 			contentStyle={ styles.contentContainer }
+			testID="editor-help-modal"
```
Test ID added as the bottom sheet modal does not provide any user-facing identifier from which to query this element. This test ID is used to trigger actions like dismissing the modal via swipe, overlay press, or hardware back button.

dcalhoun committed:
Replace unnecessary React Hook check

Defining a functioning `remove` method for the `Keyboard` mock is more
appropriate than modifying the application code for testing purposes.
The original mock did not provide a way to remove registered callbacks,
which resulted in stale callbacks updating React state erroneously.

SiobhyB commented on packages/editor/src/components/editor-help/test/index.native.js:
```diff
@@ -0,0 +1,85 @@
+	const closeMock = jest.fn();
+	const screen = render( <EditorHelp isVisible close={ closeMock } /> );
+	const closeButton = await screen.findByA11yLabel( 'Go back' );
+	fireEvent.press( closeButton );
```
While researching to understand the code here, I came across [this article](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library) that notes the following:

> [@testing-library/user-event](https://github.com/testing-library/user-event) is a package that's built on top of fireEvent, but it provides several methods that resemble the user interactions more closely. In the example above, fireEvent.change will simply trigger a single change event on the input. However the type call, will trigger keyDown, keyPress, and keyUp events for each character as well. It's much closer to the user's actual interactions. This has the benefit of working well with libraries that you may use which don't actually listen for the change event.

> We're still working on @testing-library/user-event to ensure that it delivers what it promises: firing all the same events the user would fire when performing a specific action. I don't think we're quite there yet and this is why it's not baked-into @testing-library/dom (though it may be at some point in the future). However, I'm confident enough in it to recommend you give it a look and use it's utilities over fireEvent.

I see `userEvent` being used on the web-side of Gutenberg, but not for mobile and I'm not sure if there's a specific reason for that. I wasn't able to find any resources around its use in React Native specifically, but seems like it may be worthwhile adopting?

dcalhoun replied:
> While researching to understand the code here, I came across [this article](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library)

This is a fantastic article! I had not read this. Thank you for sharing. 

> I see `userEvent` being used on the web-side of Gutenberg, but not for mobile and I'm not sure if there's a specific reason for that. I wasn't able to find any resources around its use in React Native specifically, but seems like it may be worthwhile adopting?

I had not seen `@testing-library/user-event` before. Its mission does appear to be very useful and relevant for writing better tests. That said, I do not believe we can leverage it today with our native tests, it does not appear there is support for—or even plans to support—React Native. If that support is introduced at some point, I think we should explore/consider leveraging the library.

dcalhoun replied:
I did find https://github.com/callstack/react-native-testing-library/pull/674, which attempts to make RNTL's `fireEvent.press` more aligned with the approach taken by `@testing-library/user-event`.

SiobhyB replied:
Ah, I see! I wasn't fully grokking the separation between React Testing Library and Native Testing Library. Thank you for sharing, I'll follow along with that GitHub issue and will bookmark Callstack's repository for future reference. :)

SiobhyB commented on packages/editor/src/components/editor-help/test/index.native.js:
```diff
@@ -0,0 +1,85 @@
+
+it( 'lists help topics', async () => {
+	const screen = render( <EditorHelp isVisible /> );
+	const helpTopic = await screen.findByText( 'Customize blocks' );
```
Is there a concern about this text possibly changing in the future? If so, how did you go about determining this as the most appropriate selector? I'm keen to learn more about the thought process that goes into weighing up the pros and cons of these types of decisions.

dcalhoun replied:
> Is there a concern about this text possibly changing in the future? If so, how did you go about determining this as the most appropriate selector? I'm keen to learn more about the thought process that goes into weighing up the pros and cons of these types of decisions.

[As mentioned](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-the-wrong-query) in the article you shared in your previous comment, [the recommendation](https://testing-library.com/docs/queries/about/#priority) is to query resembling how users interact with our code. However, React Native Testing Library (RNTL)—the extension that provides from for React Native—does not support every query available in the parent Testing Library (TL). So, we must adapt our practices to what is available to us. 

Following the TL guides, I think the highest priority RNTL query is `ByText`, followed by `ByA11yLabel` when there is no user-facing text. As you mentioned, querying by these does come with [the tradeoff](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-querying-by-text) that they are less "stable," i.e. small content/code changes could result in test breakages. However, I believe this tradeoff to be worthwhile and the breakage a "cheap" fix. 

In regards to _which specific Help topic I chose to query_, it was largely random. I opted to query the last Help topic. 

So, overall it is largely a judgement call, but the TL guides do a really good job of providing a "toolkit" for identifying the best query for your circumstance.

SiobhyB replied:
Thank you so much for those details and elaborating here, this type of insight is incredibly valuable for my learning and I appreciate it. 🙇‍♀️

SiobhyB reviewed (approved):
LGTM, I confirmed the new tests run and pass successfully. 🎉 

I've left a couple of questions that mainly come from my curiosity and wanting to learn, but nothing to block this PR. 

Thanks, David!

fluiddot commented on packages/editor/src/components/editor-help/test/index.native.js:
```diff
@@ -0,0 +1,85 @@
+	const helpTopic = await screen.findByText( 'Customize blocks' );
+	fireEvent.press( helpTopic );
+
+	const detailScreenTitle1 = screen.getByText(
+		'Each block has its own settings. To find them, tap on a block. Its settings will appear on the toolbar at the bottom of the screen.'
+	);
+	expect( detailScreenTitle1 ).toBeTruthy();
```
I'm wondering if we could remove this part as it has been already tested in the previous test `navigates to help topic detail screen`, wdyt?

dcalhoun replied:
Addressed in https://github.com/WordPress/gutenberg/pull/34431/commits/2f317c49fb2645af9789e66f8445a986c50a342b.

fluiddot commented on packages/editor/src/components/editor-help/test/index.native.js:
```diff
@@ -0,0 +1,85 @@
+	const backButton = screen.getAllByA11yLabel( 'Go back' );
+	fireEvent.press( backButton[ backButton.length - 1 ] );
+
+	// Currently logs `act` warning due to https://git.io/JYYGE
+	await waitForElementToBeRemoved( () =>
+		screen.getByText(
+			'Each block has its own settings. To find them, tap on a block. Its settings will appear on the toolbar at the bottom of the screen.'
+		)
+	);
+} );
```
Probably you could use the `waitFor` helper we used on other integration tests to prevent the warning:
https://github.com/WordPress/gutenberg/blob/8e4ebfae09351aa5d21fe072deb16d265bdb130e/test/native/helpers.js#L67-L101

dcalhoun replied:
The challenge with this particular assertion is that we are asserting its value is `null`. If we... 

```js
await waitFor(() => elementCurrentlyVisibleButShouldGoAway)
```

...the test continues forward immediately (and fails). If we...

```js
await waitFor(() => expect().toBeNull())
```

...the test never passes because I believe `toBeNull` doesn't return anything. 

I believe this is likely a shortcoming of our internal `waitFor` implementation, where we [expect a truthy value](https://github.com/WordPress/gutenberg/blob/acf909f8f4a15db031fb5dce3979d7f8556f4162/test/native/helpers.js#L79). So, there is likely room to implement a custom `waitForElementToBeRemoved` or refactor our `waitFor` to not check the presence of a truthy value. E.g. it would appear RNTL's `waitFor` merely [ensures the expectation does not throw an error](https://github.com/callstack/react-native-testing-library/blob/54d970c94013966c8839ee318b99737d64296d83/src/waitFor.js#L138-L167). 

However, I plan to merge this work as-is as I am limited on time currently. We can hopefully improve the error log in this test at a later point. Let me know if you think my understanding of our internal `waitFor` makes sense.

fluiddot replied:
Thanks for clarifying it ❤️ ! Interesting, I didn't realize about this particularity of asserting `null`.

> I believe this is likely a shortcoming of our internal waitFor implementation, where we expect a truthy value. So, there is likely room to implement a custom waitForElementToBeRemoved or refactor our waitFor to not check the presence of a truthy value. E.g. it would appear RNTL's waitFor merely ensures the expectation does not throw an error.

Right, `waitFor` won't work for this case as it's always expecting the result to be truthy. I think this behavior makes sense since it's expected that the callback passed to the function returns a value. I'd lean towards implementing a custom `waitForElementToBeRemoved` which probably would reuse most of the code we have for `waitFor`.

> However, I plan to merge this work as-is as I am limited on time currently. We can hopefully improve the error log in this test at a later point. Let me know if you think my understanding of our internal waitFor makes sense.

Agreed, the purpose of this comment was just to capture insights about the use of `waitFor`, so please feel free to merge it :shipit:.

fluiddot commented on packages/editor/src/components/editor-help/test/index.native.js:
```diff
@@ -0,0 +1,85 @@
+	const closeButton = await screen.findByA11yLabel( 'Go back' );
+	fireEvent.press( closeButton );
+
+	expect( closeMock ).toHaveBeenCalled();
```
I was wondering if we should also verify that the modal is actually dismissed but I realized that the `BottomSheet` component is always rendered, so there's no easy way to determine this unless we check the `isVisible` prop.

dcalhoun replied:
Possibly. Yes, `BottomSheet`'s persistent render might make it difficult. Also, I might recommend that we test that specific functionality with a test targeting `BottomSheet` instead. `BottomSheet` already is already fairly complex, so I believe it likely warrants its own tests.

fluiddot replied:
> Also, I might recommend that we test that specific functionality with a test targeting BottomSheet instead. BottomSheet already is already fairly complex, so I believe it likely warrants its own tests.

Totally agreed, `BottomSheet` component is big and complex enough to deserve its own tests.

fluiddot reviewed (approved):
Awesome job @dcalhoun 🎊 !

The PR looks great to me :shipit:, I've just posted a couple of suggestions but nothing critical, feel free to skip them.

dcalhoun committed:
Remove redundant assertion

A sibling test asserts the presence of this text. The presence of the
back button also provides similar coverage

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into chore/add-editor-onboarding-tests

dcalhoun committed:
Use npm install instead of npm ci

Temporarily change install method to circumvent erroneous CI failures.

dcalhoun committed:
Revert "Use npm install instead of npm ci"

This reverts commit ba00594639fefeee5852380a0b6bf24fffa683f4.
